core/src/test/java/org/infinispan/statetransfer/CommitManagerTest.java

```diff
@@ -0,0 +1,192 @@
+
+      // Recently create manager is not tracking anything.
+      assertFalse(manager.isTracking(Flag.PUT_FOR_STATE_TRANSFER));
+      assertFalse(manager.isTracking(Flag.PUT_FOR_X_SITE_STATE_TRANSFER));
+
+      // Start tracking state transfer.
+      manager.startTrack(Flag.PUT_FOR_STATE_TRANSFER);
+      assertTrue(manager.isTracking(Flag.PUT_FOR_STATE_TRANSFER));
+
+      // Stop tracking state transfer.
+      manager.stopTrack(Flag.PUT_FOR_STATE_TRANSFER);
+      assertFalse(manager.isTracking(Flag.PUT_FOR_STATE_TRANSFER));
+
+      // Start tracking cross site state transfer.
+      manager.startTrack(Flag.PUT_FOR_X_SITE_STATE_TRANSFER);
+      assertTrue(manager.isTracking(Flag.PUT_FOR_X_SITE_STATE_TRANSFER));
+
+      // Stop tracking cross site state transfer.
+      manager.stopTrack(Flag.PUT_FOR_X_SITE_STATE_TRANSFER);
+      assertFalse(manager.isTracking(Flag.PUT_FOR_X_SITE_STATE_TRANSFER));
+   }
+
+   public void shouldCommitEntriesForSegment() throws InterruptedException {
+      final int numberOfSegments = 10;
+
+      final InternalDataContainer container = mock(InternalDataContainer.class);
+      final PersistenceManager persistenceManager = mock(PersistenceManager.class);
+      final TimeService timeService = mock(TimeService.class);
+
+      final CommitManager manager = new CommitManager();
+      TestingUtil.inject(manager, container, persistenceManager, timeService);
+
+      // Start tracking for state transfer.
+      manager.startTrack(Flag.PUT_FOR_STATE_TRANSFER);
+
+      // Create some entries associated with segments.
+      Map<Integer, Collection<CacheEntry>> segments = new HashMap<>();
+      for (int i = 0; i < numberOfSegments; i++) {
+         List<CacheEntry> entries = new ArrayList<>();
+         for (int j = 0; j < 10; j++) {
+            String formatted = String.format("value-%d-%d", i, j);
+            entries.add(new ReadCommittedEntry(formatted, formatted, null));
+         }
+         segments.put(i, entries);
+      }
+
+      // Concurrently commit all entries from all segments.
+      final CountDownLatch barrier = new CountDownLatch(1);
+      List<CompletionStage<Void>> futures = new ArrayList<>();
+      for (Map.Entry<Integer, Collection<CacheEntry>> segment : segments.entrySet()) {
+         for (CacheEntry cacheEntry : segment.getValue()) {
+            futures.add(manager.commit(cacheEntry, Flag.PUT_FOR_STATE_TRANSFER, segment.getKey(), false, null));
+         }
+      }
+
+      // Await for all entries to be committed.
+      CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+            .whenComplete((v, t) -> {
+               if (t != null) {
+                  throw new TestException(t);
+               }
+               barrier.countDown();
+            });
+
+      if (!barrier.await(10, TimeUnit.SECONDS)) {
+         throw new TestException("Timed out waiting for all entries to be committed.");
+      }
```

Comment:
The `CompletableFuture` from `CommitManager.commit()` should be done since you are mocking all the components. 
All this code can be replaced with a dual for loop:
for (int segment = 0; ...)
  for (int i ; i < number of keyes)
     manager.commit().get(1, seconds)